Add github version range support

computmaxer opened the pull request:
_I know this is not the proper way to submit contributions, but I want to get feedback from dart-lang folks and make sure it is something that could potentially be accepted before I spend any more time on this._

This is a fix for #1250. I work for Workiva (met some of you at the Dart Summit!) and we now have quite a few dart libraries in our Github organization. Some of these are public and OSS, but many of them cannot be. #1250 is a big issue for us for the private libraries as it requires us to update pubspecs all over the place anytime a patch version of some library is released, and consumers end up having a big list of dependency overrides while those transitions are in progress.

This PR implements version ranges for github dependencies with the following in a `pubspec.yaml`:

```yaml
dependencies:
  w_flux:
    version: "^1.0.0"
    git: git@github.com:Workiva/w_flux.git
```

It assumes that versions will exist as git tags on the repo, and uses `git tag -l` to list the tags. It then collects a list of the tags that are considered valid version strings using `Version.parse()`. 

The biggest issue I have had so far is dealing with git tags that include the "v" in the name, i.e. "v1.0.0". I can strip the "v" before giving it to the version solver, but then when it tries to get the ref (`_getRef`) it of course cannot find it without the "v" that was previously removed. At Workiva we do not use the preceding "v" in our tag names, but it seems like many people do so I am assuming this will be a big issue for this feature.

It seems like it would be possible for the Version class to accept "v" versions strings in its `parse` constructor, and keep track of it so that it can put it back when `toString()` is called. This would require a change to pub_semver, though.

I'm hoping to get some feedback from the Dart team on whether or not we should continue down this path. Please let me know your thoughts! 

@nex3 @kevmoo @munificent

nex3 commented:
I don't think you need to do that many backflips to get "v" versions working. Just add logic to `GitSource._getEffectiveRef` to check for refs named `"v$version"` and `"$version"`. That way you can have everything else just track the version number as part of the package ID, like it does for hosted dependencies.

computmaxer commented:
Sorry for the delay! Finally had some time to spend on this yesterday. I've implemented @nex3 's suggestion for dealing with "v" version tags.

I also fixed a major issue with my initial approach. I was creating bare-bones `Pubspec` objects with just `name` and `version` in `getVersions()`, but it turns out those pubspecs need to be complete since one will eventually be used to determine which nested dependencies to get. So unfortunately we'll need to return the entire pubspec for each potential valid version. I have implemented that here by calling `describeUncached()` with each valid version string in a `PackageId`. This has the potential to impact performance since that will checkout each tag and load a pubspec, serially.

To help combat that issue I added a flag to enable this feature per dependency. Looks like this:

```yaml
dependencies:
  w_flux:
    version: "^1.0.0"
    git: 
      url: git@github.com:Workiva/w_flux.git
      use_version_tags: true
```

Pub will only `getVersions()` for a particular git dependency if `use_version_tags` is true. This has the additional benefit of insulating this feature from existing tests and functionality, reducing its impact.

---

All unit tests are passing and I have tested this with a number of repositories, with and without "v"s in the version tags. Let me know what you think and what the next steps would be if you want to move forward with this solution.

greglittlefield-wf commented:
If checking out each tag is slow, would using [`git show`](https://git-scm.com/docs/git-show) (e.g., `git show TAGNAME:pubspec.yaml`) instead of `git checkout` help to significantly reduce that performance hit?

computmaxer commented:
@greglittlefield-wf nice! Didn't know about `git show`. I've added a commit that uses that instead of `describeUncached()`.

kevmoo commented:
@computmaxer If you're fixing your pull request, please rebase against master and do a `--force` push to your staging branch. This makes the pull request "clean".

kevmoo commented on lib/src/source/git.dart:
```diff
@@ -40,6 +41,49 @@ class GitSource extends CachedSource {
+      } on FormatException {}
+    }
+
+    if (validVersions.length > 0) {
```
use `.isNotEmpty`

kevmoo commented on lib/src/source/git.dart:
```diff
@@ -254,8 +303,19 @@ class GitSource extends CachedSource {
   Future<String> _getRev(PackageId id) {
-    return git.run(["rev-list", "--max-count=1", _getEffectiveRef(id)],
-        workingDir: _repoCachePath(id)).then((result) => result.first);
+    var ref = _getEffectiveRef(id);
```
Consider making this method `async`

kevmoo commented on lib/src/source/git.dart:
```diff
@@ -285,8 +345,28 @@ class GitSource extends CachedSource {
   Future _checkOut(String repoPath, String ref) {
-    return git.run(["checkout", ref], workingDir: repoPath).then(
-        (result) => null);
+    return git.run(["checkout", ref], workingDir: repoPath)
```
ditto: consider making async

kevmoo commented on lib/src/source/git.dart:
```diff
@@ -285,8 +345,28 @@ class GitSource extends CachedSource {
+  /// It is possible that a pubspec didn't always exist, return null if
+  /// that is the case.
+  Future<Pubspec> _getPubspec(String repoPath, String ref) {
+    return git.run(['show', '$ref:pubspec.yaml'], workingDir: repoPath)
```
ditto: consider async

computmaxer commented:
@kevmoo comments addressed. Would you all commits squashed into one as well?

kevmoo commented on lib/src/source/git.dart:
```diff
@@ -253,9 +302,20 @@ class GitSource extends CachedSource {
+    var ref = _getEffectiveRef(id);
+    return git
+        .run(["rev-list", "--max-count=1", ref], workingDir: _repoCachePath(id))
+        .then((result) => result.first)
```
By async, I also meant getting red of `.then` and `.catchError` and using `await`, too. :smile:

computmaxer replied:
Ok. Do you guys _always_ prefer async/await over future chaining, then?

kevmoo replied:
In general, yes. It's easier to read and our debugger actually has special support for async/await – that we don't get if you use the Future API directly.

@nex3 has found a few cases where it's better to use the API directly, but she'd have to comment.

kevmoo commented on lib/src/source/git.dart:
```diff
@@ -284,9 +344,29 @@ class GitSource extends CachedSource {
-  Future _checkOut(String repoPath, String ref) {
-    return git.run(["checkout", ref], workingDir: repoPath).then(
-        (result) => null);
+  Future _checkOut(String repoPath, String ref) async {
```
ditto

computmaxer replied:
Fixed.

nex3 replied:
Note that you won't need to modify this if you change `downloadToSystemCache` to use `_getRef`.

kevmoo commented:
@computmaxer I'd wait to squash until reviews are done – but it's good to do once you think you're done

kevmoo commented on lib/src/source/git.dart:
```diff
@@ -253,9 +302,22 @@ class GitSource extends CachedSource {
+            workingDir: _repoCachePath(id));
+        return result.first;
+      }
+      throw e;
```
Use `rethrow` – it preserves the stack trace

computmaxer replied:
nice :+1:

nex3 commented on lib/src/source/git.dart:
```diff
@@ -40,6 +41,49 @@ class GitSource extends CachedSource {
+  /// Anything that Version.parse understands is considered a version,
+  /// it will also attempt to strip off a preceding 'v' e.g. v1.0.0
+  Future<List<Pubspec>> getVersions(String name, description) async {
+    if (!_useVersionTags(description)) {
```
I'm really not a fan of requiring users to explicitly annotate to get this behavior. For the most part, I don't expect people to be using version declarations for Git dependencies at all unless they want some sort of resolution, so requiring them to write more stuff seems annoying. And now that we're using `git show`, it shouldn't be too bad to get all the pubspecs.

computmaxer replied:
The issue I'm having here is that I can't seem to access the `version` in this method, as it is not part of the description.

```
w_transport:
  version: ^1.0.0
  git: git@github.com:Worvia/w_transport.git
```
> description --> "git@github.com:Workiva/w_transport.git"

```
w_transport:
  version: ^1.0.0
  git: 
    url: git@github.com:Worvia/w_transport.git
    use_version_tags: true
```
> description --> {'url': 'git@github.com:Worvia/w_transport.git', 'use_version_tags': true}

Any idea how to short-circuit based on the `version`?

greglittlefield-wf replied:
It'd be possible to access `version` if it was in the description under `git`, like you initially had it, right?
```yaml
dependencies:
  w_flux:
    version: "^1.0.0"
    git: git@github.com:Workiva/w_flux.git
```

Is there any reason that version can't be under `git`?

computmaxer replied:
The version solver (which is abstracted from the package source) won't respect it when it is under `git`.

nex3 replied:
Yikes, good point. I'd just change `Source.getVersions()` to take a `PackageRef` rather than a name and description, and check whether the version constraint is `any` or not.

computmaxer replied:
Do you mean `PackageDep` instead of `PackageRef`? `PackageRef` does not have a `VersionConstraint` available.

nex3 commented on lib/src/source/git.dart:
```diff
@@ -40,6 +41,49 @@ class GitSource extends CachedSource {
+    }
+
+    PackageId id =
+        new PackageId(name, this.name, null, _getDescription(description));
```
Passing `null` for the version here is a smell. It looks like `_repoCachePath` doesn't actually require a version, so it should be modified to take a `PackageRef` instead of a `PackageId`. You can update other call sites by calling `PackageId.toRef()` to get the ref.

computmaxer replied:
Addressed this in https://github.com/computmaxer/pub/commit/1a9120c4f67baf65f026c98fd282b4bfb9d0499b

nex3 commented on lib/src/source/git.dart:
```diff
@@ -40,6 +41,49 @@ class GitSource extends CachedSource {
+    if (!entryExists(cachePath)) {
+      // Must have the repo cloned in order to list its tags
+      await _clone(_getUrl(id), cachePath);
+    }
```
This is the same logic as in `_ensureRevision`; consider factoring it out into `String _ensureRepo(PackageRef ref)`.

computmaxer replied:
Addressed in https://github.com/dart-lang/pub/commit/3d9c0abe7d8e774e1f7aca1765e9cf6e0ef17891

nex3 commented on lib/src/source/git.dart:
```diff
@@ -40,6 +41,49 @@ class GitSource extends CachedSource {
+      await _clone(_getUrl(id), cachePath);
+    }
+
+    List results = await git.run(["tag", "-l"], workingDir: cachePath);
```
If you pass a pattern to `--list`, you can cut down on the number of irrelevant tags you have to look at. The pattern `*.*.*` will match all valid semantic versions with or without the leading `v` (although obviously it will also match a lot of other stuff).

computmaxer replied:
Addressed in https://github.com/computmaxer/pub/commit/59f2e6f28fdad7bdbe272a72e78f532c9ad27537

nex3 commented on lib/src/source/git.dart:
```diff
@@ -40,6 +41,49 @@ class GitSource extends CachedSource {
+
+    List results = await git.run(["tag", "-l"], workingDir: cachePath);
+    List<Pubspec> validVersions = [];
+    for (String version in results) {
```
Consider using `Iterable.where` here.

computmaxer replied:
This would require an additional loop - the `where` which selects the valid version strings, and another that loops through the valid version strings and awaits the pubspec for each.

I sort of prefer it as one loop, but if you would rather have that separation I can change it.

nex3 commented on lib/src/source/git.dart:
```diff
@@ -40,6 +41,49 @@ class GitSource extends CachedSource {
+        // Use Version.parse to determine valid version tags
+        Version validVersion = new Version.parse(version);
+        // Fetch the pubspec for this version
+        Pubspec pubspec = await _getPubspec(cachePath, validVersion.toString());
```
`validVersion.toString()` here will break if the actual tag started with a `v`.

computmaxer replied:
Addressed in https://github.com/dart-lang/pub/commit/8b1aabdd4396e8f2b7147d4b7f4fc8f4f94eda9d

nex3 commented on lib/src/source/git.dart:
```diff
@@ -40,6 +41,49 @@ class GitSource extends CachedSource {
+        if (pubspec != null) {
+          validVersions.add(pubspec);
+        }
+      } on FormatException {}
```
We should definitely catch format exceptions from the version parse, but I don't think we should catch errors caused by an invalid pubpsec. It's a pretty safe assumption that once we're here, the user intentionally opted into version resolution over a given repo, and having some versions silently not appear may be confusing.

computmaxer replied:
Consider the case where a repo initially was tagged 1.0.0 without a pubspec (perhaps it used to be in JS or something ¯\_(ツ)_/¯ ). But all tags after 2.0.0 have a pubspec.
If we don't handle invalid pubspec here, this will cause an exception attempting to load the pubspecs for versions <2.0.0 and they won't be able to use version resolution at all for that given repo.

nex3 replied:
I think with my refactor, we'll be able to avoid trying to load pubspecs until they're actually required by the version solver, which should mitigate this.

nex3 commented on lib/src/source/git.dart:
```diff
@@ -40,6 +41,49 @@ class GitSource extends CachedSource {
+    for (String version in results) {
+      // Strip preceding 'v' character so 'v1.0.0' can be parsed into a Version
+      if (version.startsWith('v')) {
+        version = version.substring(1);
```
This should do some de-duplication to ensure that, if a version exists both with and without `v`, only the canonical one (that is, the one without `v`) is returned.

computmaxer replied:
Addressed here: https://github.com/dart-lang/pub/commit/c9353549715b634bd59a9445c462ee612985be5d

nex3 commented on lib/src/source/git.dart:
```diff
@@ -253,9 +302,22 @@ class GitSource extends CachedSource {
-  Future<String> _getRev(PackageId id) {
-    return git.run(["rev-list", "--max-count=1", _getEffectiveRef(id)],
-        workingDir: _repoCachePath(id)).then((result) => result.first);
+  Future<String> _getRev(PackageId id) async {
```
If you're changing `_getRev` rather than `_getEffectiveRef` (which makes sense), you'll need to update `downloadToSystemCache` to use the former so that it behaves properly with `v`-versions.

computmaxer replied:
Addressed in https://github.com/dart-lang/pub/commit/01e6787b1bd9bdc5fec877ee288241810a747da4

kevmoo replied:
Thanks for linking to the commit w/ the fix. Makes these reviews so much better. :smile:

nex3 commented on lib/src/source/git.dart:
```diff
@@ -253,9 +302,22 @@ class GitSource extends CachedSource {
-    return git.run(["rev-list", "--max-count=1", _getEffectiveRef(id)],
-        workingDir: _repoCachePath(id)).then((result) => result.first);
+  Future<String> _getRev(PackageId id) async {
+    var ref = _getEffectiveRef(id);
```
`_getEffectiveRef` doesn't really make much sense in the new world. It returns `"HEAD"` if the ID doesn't specify a revision, but that's incorrect if version resolution needs to be done. I think what you want to do here is more like:

```dart
Future<String> _getRev(PackageId id) {
  var description = _getDescription(id);

  // If the id is already resolved, there's no need to rev-list to look for it.
  var resolved = description['resolved-ref'];
  if (description is Map && resolved != null) return resolved;

  // If [id] includes a version and not a ref, try "rev-list $version" and
  // "rev-list v$version". Otherwise, try "rev-list ${_getRef(id) ?? 'HEAD'}".
}
```

computmaxer replied:
I'm having an issue with this. Here's what the code roughly looks like:

```dart
Future<String> _getRev(PackageId id) {
  print('_getRev called: ${id.toString()}');
  ...

  var ref = description is Map ? description['ref'] : null;
  if (ref == null && id.version != null && id.version != Version.none) {
    print('Trying to use version: ${id.version.toString()}');
   // return the rev-list $version result
  } else {
    print('Returning _getRef(id) or HEAD');
    // return rev-list _getRef(id) ?? 'HEAD';
  }
}
```

For some reason `id.version` ends up getting set to something even if no `version` is specified in the pubspec. So if I put some print statements in this `_getRev` method the output looks like this when I run a simple test like `test/get/git/check_out_test.dart`

```
  | MSG : Resolving dependencies...
  | _getRev called: foo 0.0.0 from git
  | Returning _getRef(id) or HEAD
  | _getRev called: foo 0.0.0 from git
  | Returning _getRef(id) or HEAD
  | _getRev called: foo 0.0.0 from git
  | Returning _getRef(id) or HEAD
  | _getRev called: foo 0.0.0 from git
  | Returning _getRef(id) or HEAD
  | _getRev called: foo 1.0.0 from git
  | Trying to use version: 1.0.0
```
That's with a description of `foo: "../foo.git"` (no version specified). So by the logic in the if condition, it will try to get a revision with `1.0.0` which fails as it should. It should have hit the `else` case instead for this to work properly.

If I leverage `_useVersionTags(id)` in that if condition I can correctly determine whether or not the version should be used as a ref. This is what the existing code in `_getEffectiveRef` is doing. However, you mentioned you don't want to rely on an additional annotation for this to work.

I will keep looking for a solution, but let me know if you have any ideas that will solve this. Then I should be able to get rid of the `_useVersionTags(id)` stuff.

computmaxer replied:
`id.version` ends up being set to the version number in the pubspec.yaml after the solver has selected a pubspec: https://github.com/dart-lang/pub/blob/master/lib/src/solver/version_solver.dart#L257

`resolveId` is called once again after a solution has been reached: https://github.com/dart-lang/pub/blob/master/lib/src/solver/backtracking_solver.dart#L168. `resolveId` uses `_ensureRevision` which uses `_getRev`.

At this point `id.version` has something even if one wasn't specified in the consumer's pubspec.

So this is why `if (ref == null && id.version != null && id.version != Version.none)` is not a sufficient condition to determine whether or not we should do a `rev-list $version`.

Open to suggestions :)

nex3 replied:
Yeah, I forgot that `PackageId` doesn't include the original version constraint because it's not part of the description. This is going to be trickier than I thought.

The underlying issue is that Git IDs are no longer context-independent in the way they used to be; we don't know whether one was created due to new-style version resolution or old-style looking at `HEAD`. Git does have a means of "resolving" package IDs, but because `getVersions()` returns `Pubspec`s rather than `PackageId`s, those resolved IDs don't make their way into the solver.

This probably calls for a full refactor of `getVersions()` that changes the signature to `List<PackageId> getVersions(PackageRef ref)` and moves some caching logic into the hosted source. This is out of the scope of what you're doing, though, so I'll implement it myself; hold tight until then.

nex3 commented:
@computmaxer Done reviewing!

computmaxer commented:
@nex3 I've addressed each of your comments, some with further questions.

computmaxer commented:
@greglittlefield-wf @travissanderson-wf @trentgrover-wf @evanweible-wf Perhaps you guys have some ideas for some of these remaining issues?

nex3 commented:
I'm really really sorry for pulling the rug out from this patch at the last minute, but the more I think about its UI and the more I talk to @rnystrom, the less I think we'll be able to get it to work. On one hand, I don't think we're going to be able to trigger this based on a version constraint existing; that information just doesn't get piped where it would need to go, and it doesn't make sense in general to add that piping. But on the other hand, we really don't want to have to hide this behind a cumbersome flag that would have to go on every dependency. So I don't think it's going to make it. Again, I'm sorry I didn't think this through thoroughly earlier on.

trentgrover-wf commented:
The main problem that prompted this effort is that there doesn't seem to be a complete and secure mechanism for version solving with private Dart packages via pub.

The 2 main mechanisms for consuming dependencies are:

1) Pub server

The default pub server works to solve dependency constraints perfectly for public packages only.  Using this functionality with private dart libraries requires standing up and maintaining a private pub server.  There are also some functional gaps in the existing implementation:
  - There's no auth mechanism for pub get (issue raised and closed here: https://github.com/dart-lang/pub/issues/1327).
  - Setting the PUB_HOSTED_URL environment variable doesn't automatically fall back to the normal pub server if the custom server doesn't contain the package requested (mentioned here: https://github.com/dart-lang/pub/issues/1241).
  - Using the full form hosted dependency syntax would resolve the PUB_HOSTED_URL fallback issue, but requires more cumbersome per dependency syntax.

2) GitHub repos

The existing GitHub repo functionality in pub supports private pub packages well in a secure way, but contains no version constraint solving (hence this attempt at a solution), which makes dependency management across complex dependency trees extremely cumbersome.
  - Given the two avenues available to us, it seems easier and more useful to a wide range of consumers if we can get the constraint solving of GitHub repos implemented (auth already works, GitHub can provide the infrastructure / server, and fallback to existing public pub is transparent and automatic).
  - Even hiding the GitHub range support behind some cumbersome per-dependency flag seems no worse than the existing full form hosted dependency syntax.

If there are significant reasons why GitHub version solving isn't technically feasible, it would be really helpful to have a more straightforward way to accomplish the same goals via private pub packages.  Is there anything we can do to get something workable via 1 of these 2 paths?

kevmoo commented:
@travissanderson-wf @trentgrover-wf I'll chat with @nex3 to see what our options are – reading https://github.com/dart-lang/pub/issues/1327, it sounds like you'd be open to running your own server

computmaxer commented:
@kevmoo we are open to it, but would rather have support for Git version solving to avoid maintenance headaches and the auth hurdle. 

> I don't think we're going to be able to trigger this based on a version constraint existing; that information just doesn't get piped where it would need to go, and it doesn't make sense in general to add that piping.

Why doesn't it make sense to add that piping? It seems like `getVersions` would just need to be given a `PackageDep` instead of a `PackageRef`. Looking briefly at the code, it seems like that might not be a trivial change (may affect a lot of the version solver code), but wouldn't it make sense to have constraint information when solving the version? How does that become a blocker for this feature? Pub already has the design/architecture to support version solving across different package sources, but only hosted packages are actually able to use it. Doesn't it make sense to refactor that functionality slightly to allow the design to be utilized?

trentgrover-wf commented:
@kevmoo any further thoughts on this issue?

nex3 commented:
If you're looking to only allow authenticated downloads, I think your best bet is to either run some sort of VPN or to use HTTP basic auth and have people include their credentials in the `PUB_HOSTED_URL` variable. Either way you'll also need to set up your server to fall back on `pub.dartlang.org` packages.

trentgrover-wf commented:
So is there no chance of a GitHub based solution (eliminates a bunch of infrastructure requirements)?

If you could point us to your specific areas of concern, we're more than happy to devote some dev time to it, but it's sounding like the NO is hardening...

nex3 commented:
I don't think we're going to find a way to make version range resolution work with GitHub, no.

johncblandii commented:
That kind of sucks to go through so much PR review/work to say "no" to the whole idea of it later.

The feature is useful. Hope you reconsider @nex3.

kevmoo commented:
@johncblandii It absolutely sucks. We're very sorry.

We try **very** hard to make sure these types of changes don't have long-term implications or side-effects that can hurt the ecosystem.

@nex3 is especially diligent when it comes to these types of changes – although, as she mentioned, we realized there were gnarly edge cases we didn't initially consider when we first encouraged the work to proceed.

We'd be more than happy to discuss other approaches to this problem – we know path dependencies are a pain.

johnryan-wf commented:
In hindsight, knowing this was a dead end would have allowed us to work on an internal pub server.  

We want to make sure that whatever solution we end up with is solid, but I'm not sure why this can't be accomplished (in principle.)  It *seems* like an elegant way to support private dependencies.  So close :)

kevmoo commented:
I'll chat w/ @nex3 a bit more tomorrow. We know git (and GitHub) is a really close fit here.

Maybe we can get creative.

trentgrover-wf commented:
@kevmoo @nex3 any further update?  sorry to badger you

munificent commented:
Various holidays and related vacations are kicking in, so @kevmoo and @nex3 might be offline for a while, but they will follow up on this. :)

trentgrover-wf commented:
:+1:

trentgrover-wf commented:
Now that the holidays have passed, any further update @kevmoo @nex3?  We're trying to nail down how we want to address this issue internally, so your answer will help us determine what resources to devote to it.

trentgrover-wf commented:
For posterity, here's where we landed:
- Google will not be merging GitHub version range support because of the issues noted above and because the risk associated with introducing breaking changes this low in the stack is too high.
- We're going with the private pub server route:
  - Google has committed to adding some oauth capability to pub so that `pub get` can be performed in a secure way without requiring the auth credentials to reside in the pubspec (as would be required if using only HTTP basic auth).  Delivery date TBD.
  - Once we have a private pub server up and running with oauth, we'll use the hosted package syntax with a custom hosted url specified for each of our private pubspec dependencies.

munificent commented:
Sounds right to me. Just for the record (and you probably already know this):

> * Once we have a private pub server up and running with oauth, we'll use the hosted package syntax with a custom hosted url specified for each of our private pubspec dependencies.

This is supported already in pub, so there's no work on our end to do. It looks like:

```yaml
dependencies:
  transmogrify:
    hosted:
      name: transmogrify
      url: http://your-package-server.com
    version: ^1.4.0
```

computmaxer commented:
@kevmoo @munificent @nex3 
We planned on firewalling off our private pub server from the internet and only allowing our office/VPN IPs until the OAuth2 support for `pub get` is added.

I am having trouble getting this working with the Managed VM GAE deployment. I've added the firewall rule to allow our office IPs in Networking --> Firewall rules, but I can't seem to block access to everything else. I believe this would work if it was just a GCE deployment, but it seems GAE must add some rules or routes to allow traffic to be routed from the internet to the instances.

Do you guys know if it is possible to implement an IP whitelist for a GAE Managed VMs deployment?

nex3 commented:
I'm far from an expert on AppEngine infrastructure, and I think the same is true for Kevin and Bob. Maybe @mkustermann would know; he works on pub.dartlang.org.

kevmoo commented:
Try https://groups.google.com/forum/#!forum/app-engine-managed-vms

This is purely outside of the Dart team

computmaxer commented:
@nex3 @kevmoo 
Do you all have any updates on the additional OAuth2 functionality for pub? We are yet to have a functioning private pub server approved for production use (I'll spare you the details :smile:), and since we are approaching the end of Q1 I figured I'd check in to see where this is at. Thanks!

nex3 commented:
I haven't had a chance to work on this yet. I'll try to get it in the works next quarter.
